src/daft-plan/src/logical_optimization/rules/eliminate_cross_join.rs

```diff
@@ -0,0 +1,718 @@
+                }
+                _ => Ok(Transformed::yes(left)),
+            }
+        }
+    }
+}
+
+fn rewrite_children(
+    optimizer: &impl OptimizerRule,
+    plan: Arc<LogicalPlan>,
+) -> DaftResult<Transformed<Arc<LogicalPlan>>> {
+    plan.map_children(|input| optimizer.try_optimize(input))
+}
+
+fn flatten_join_inputs(
+    plan: LogicalPlan,
+    possible_join_keys: &mut JoinKeySet,
+    all_inputs: &mut Vec<LogicalPlanRef>,
+) -> DaftResult<()> {
+    match plan {
+        LogicalPlan::Join(join) if join.join_type == JoinType::Inner => {
+            let keys = join.left_on.into_iter().zip(join.right_on);
+            possible_join_keys.insert_all_owned(keys);
+            flatten_join_inputs(
+                Arc::unwrap_or_clone(join.left),
+                possible_join_keys,
+                all_inputs,
+            )?;
+            flatten_join_inputs(
+                Arc::unwrap_or_clone(join.right),
+                possible_join_keys,
+                all_inputs,
+            )?;
+        }
+        _ => {
+            all_inputs.push(Arc::new(plan));
+        }
+    };
+    Ok(())
+}
+
+/// Returns true if the plan is a Join or Cross join could be flattened with
+/// `flatten_join_inputs`
+///
+/// Must stay in sync with `flatten_join_inputs`
+fn can_flatten_join_inputs(plan: &LogicalPlan) -> bool {
+    // can only flatten inner / cross joins
+    match plan {
+        LogicalPlan::Join(join) if join.join_type == JoinType::Inner => {}
+        _ => return false,
+    };
+
+    for child in plan.children() {
+        if matches!(
+            child,
+            LogicalPlan::Join(Join {
+                join_strategy: None,
+                join_type: JoinType::Inner,
+                ..
+            })
+        ) && !can_flatten_join_inputs(child)
+        {
+            return false;
+        }
+    }
+    true
+}
+
+/// Extract join keys from a WHERE clause
+fn extract_possible_join_keys(expr: &Expr, join_keys: &mut JoinKeySet) {
+    if let Expr::BinaryOp { left, op, right } = expr {
+        match op {
+            Operator::Eq => {
+                // insert handles ensuring  we don't add the same Join keys multiple times
+                join_keys.insert(left, right);
+            }
+            Operator::And => {
+                extract_possible_join_keys(left, join_keys);
+                extract_possible_join_keys(right, join_keys);
+            }
+            // Fix for join predicates from inside of OR expr also pulled up properly.
+            Operator::Or => {
+                let mut left_join_keys = JoinKeySet::new();
+                let mut right_join_keys = JoinKeySet::new();
+
+                extract_possible_join_keys(left, &mut left_join_keys);
+                extract_possible_join_keys(right, &mut right_join_keys);
+
+                join_keys.insert_intersection(&left_join_keys, &right_join_keys);
+            }
+            _ => (),
+        };
+    }
```

Comment:
There's similar logic to this in other parts of the code already, I believe the only difference is this does not pull out common expressions from two sides of an or, but if we want that, we should probably add it to the same place so other optimizer rules can take advantage as well. 
https://github.com/Eventual-Inc/Daft/blob/52289308cc4d27b6f2496b2c302493d7903f6bd4/src/daft-dsl/src/optimization.rs#L58